core/state/history.go

```diff
@@ -0,0 +1,101 @@
+package state
+
+import (
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/core/trie2"
+	"github.com/NethermindEth/juno/db"
+)
+
+var _ StateReader = &StateHistory{}
+
+type StateHistory struct {
+	blockNum uint64
+	state    *State
+}
+
+func NewStateHistory(blockNum uint64, stateRoot felt.Felt, db *StateDB) (*StateHistory, error) {
```

Comment:
It looks to me that `*StateHistory` can be returned by value, also `stateRoot` can be passed by reference.